### PR TITLE
Add Ramda

### LINKED ISSUE #2312: Use Ramda (or similar) to replace general utility functions and to encourage functional programming
I propose using a library to replace some of our general utility functions e.g. `compose`, `deepGet`, `deepClone` and provide us with many more useful utilities. Then our own utilitity functions would be more specific to our use cases. I propose using Ramda because it is a library designed specifically for a functional programming style. We can also include only the parts of the Ramda library we need e.g. `import compose from 'ramda/src/compose'`.

I wouldn't want to enforce a trendy programming style with a steep learning curve which can be intimidating for beginners and newcomers but I think it's worth encouraging more use of and awareness of certain aspects of functional programming e.g. function composition, pure functions, currying, and writing code without shared state, mutable data, and side-effects. This style of coding tends to be more declarative, easier to reason about and easier to debug.

## Proposed solution

 - Create ADR for use of Ramda or similar.
 - Install Ramda
 - Replace use of general utility functions with Ramda equivalents - `compose`, `deepGet`, `deepClone`

### PR DESCRIPTION
**Pull request by ryanmccombe**
Resolves https://github.com/bbc/simorgh/issues/2312

**Overall change:** Remove three generic utility functions, and use an external dependency instead.  Less code for us to maintain

**Code changes:**

- Replace `compose` with Ramda `compose`
- Replace `deepGet` with Ramda `pathOr`
- Replace `deepClone` with Ramda `clone`

**Test notes:**
This PR has a wide impact area, but anything that failed here would very likely have been caught by automation.

Worth a general sanity check just in case.
---

- [x] I have assigned myself to this PR and the corresponding issues
- [ ] Tests added for new features
- [ ] Test engineer approval

### CODE REVIEW

File: scripts/bundleSize.sh [hunk 1]
```diff
@@ -2,8 +2,8 @@
 
 # Size limit for all bundles used by each service (K)
 # Keep these +/- 5K and update frequently!
-min=557
-max=576
+min=560
+max=580
```
Comment (ryanmccombe) [inline scripts/bundleSize.sh]: The 3 functions only push the news JS bundle size from 576 to 577 - just adding some buffer

Comment (dr3) [reply scripts/bundleSize.sh]: The buffer should be 5K from the biggest service and 5k from smallest

Comment (ryanmccombe) [reply scripts/bundleSize.sh]: Doesn't look like people are following that rule then :(  Updated the thresholds

File: scripts/bundleSize.sh [hunk 2]
```diff
@@ -2,8 +2,8 @@
 
 # Size limit for all bundles used by each service (K)
 # Keep these +/- 5K and update frequently!
-min=557
-max=576
+min=564
+max=582
```
Comment (dr3) [inline scripts/bundleSize.sh:6]: Does that mean ramda increases our bundle on news by 6kb?

Comment (dr3) [reply scripts/bundleSize.sh:6]: Ohh ok if it was wrong before nvm

**Review by dr3 — APPROVED** [review]

**Review by chris-hinds — APPROVED** [review]

**Review by jroebu14 — COMMENTED** [review]
should we maybe use `path` instead of `pathOr` as `path` has a default value of undefined.

**Comment by ryanmccombe** [comment]
> should we maybe use `path` instead of `pathOr` as `path` has a default value of undefined

Unfortunately it seems a lot of the existing uses of `deepGet` do require the value to be `null` rather than `undefined` - I believe React/JSX treats them differently when you try to render them

The `pathOr` API definitely isn't as clean - I considered leaving the existing `deepGet` as a wrapper around `pathOr` (https://github.com/bbc/simorgh/blob/aff69725e87a942afe0e1c2b34f774d83a7f4ffd/src/app/lib/utilities/deepGet/index.js) but I think it's probably better to just have fewer files / tests / code to maintain in general

People can use the cleaner `path` API in the future if the `undefined` fallback is acceptable

**Comment by jroebu14** [comment]
So absolutely none of those can just use `path`?

**Comment by ryanmccombe** [comment]
> So absolutely none of those can just use path?

The problem is I don't know which ones can and which can't, so just keeping the return value the same seems safer

The alternative is going through and analysing every instance to determine whether `null` can be replaced with `undefined` safely - the implications on the dev time and QA time would really throw the cost:benefit ratio off imo - but happy to do it if people think it's worth it!

**Comment by jroebu14** [comment]
Ah ok no worries. I thought our tests would give some level of confidence about making changes like that but you're right it is risky and would require some effort to ensure there's no bugs have crept in.

**Comment by ryanmccombe** [comment]
Around 15 tests do fail when using path, but I wouldn't have much confidence that those areas really do require `deepGet` to return explicitly `null`, or that the other places that don't have failing unit
tests wouldn't break if it no longer did return `null` as such esoteric things are rarely considered in unit tests :(

**Comment by PriyaKR** [comment]
LGTM.

**Pull request merged**